ARROW-9017: [C++][Python] Refactor scalar bindings

kszucs opened the pull request:
TODOs:
 - [x] split PRs into two, one with simplified python to arrow conversions with benchmarks
 - [x] implement union scalar on the C++ side
 - [x] store the index value for dictionary scalar
 - [x] more tests

Closes ARROW-9017, ARROW-9153

kszucs commented on python/pyarrow/tests/test_misc.py:
```diff
@@ -120,7 +120,6 @@ def test_cpu_count():
     pa.LargeListValue,
     pa.MapValue,
     pa.FixedSizeListValue,
-    pa.UnionValue,
```
Restore it

jorisvandenbossche commented on python/pyarrow/tests/test_parquet.py:
```diff
@@ -2028,7 +2028,7 @@ def test_filters_invalid_pred_op(tempdir, use_legacy_dataset):
         assert dataset.read().num_rows == 0
 
-    with pytest.raises(ValueError if use_legacy_dataset else TypeError):
+    with pytest.raises(ValueError if use_legacy_dataset else pa.ArrowInvalid):
         # dataset API returns TypeError when trying create invalid comparison
```
```suggestion
    with pytest.raises(ValueError):
```

ArrowInvalid is a ValueError, I think, so if this changed, the above should be sufficient

kszucs replied:
Nice, updated.

jorisvandenbossche commented on python/pyarrow/util.py:
```diff
@@ -41,6 +41,24 @@ def wrapper(*args, **kwargs):
+def _deprecate_class(old_name, new_class, next_version,
+                     instancecheck=True):
+    """
+    Raise warning if a deprecated class is used in an isinstance check.
```
Shouldn't it also raise when instantiated?

kszucs replied:
> Really nice!
> 
> Could we add a `is_valid` attribute to the python scalar as well? Now the only way to check for a null value is to do `.as_py() is None` ?
Definitely!
> 
> Might not be related to the changes in this PR, but reviewing this triggered me to test scalar casting:
> 
> ```
> In [2]: s = pa.scalar(pd.Timestamp("2012-01-01")) 
>    ...:  
>    ...: import pyarrow.compute as pc 
>    ...: pc.cast(s, pa.timestamp('ns'))  
> ../src/arrow/compute/kernels/scalar_cast_temporal.cc:130:  Check failed: (batch[0].kind()) == (Datum::ARRAY) 
> ...
> Aborted (core dumped)
> ```
I think the cast kernel doesn't support scalars yet, on the other hand the scalars have custom `CastTo` implementation which we may want to remove in favor of `compute.Cast`?

cc @bkietz

jorisvandenbossche commented on python/pyarrow/scalar.pxi:
```diff
@@ -16,1198 +16,704 @@
 
     def __repr__(self):
-        return 'NULL'
+        return '<pyarrow.{}: {!r}>'.format(
```
Big +1 on this better repr! Just showing the repr of the as_py value was always very confusing :)

jorisvandenbossche commented on python/pyarrow/scalar.pxi:
```diff
@@ -16,1198 +16,704 @@
     """
 
-    def as_py(self):
+    cdef array(self):
```
do we want to expose this in some form publicly as well? (like you have the "raw" `value` for timestamps)

kszucs replied:
Exposed as `values`

jorisvandenbossche commented on python/pyarrow/scalar.pxi:
```diff
@@ -16,1198 +16,704 @@
-                          ap.value_length(self.index))
-        return pyarrow_wrap_buffer(buf)
+        if sp.is_valid:
+            return {frombytes(fields[i].get().name()): Scalar.wrap(sp.value[i])
```
Should we directly take the `as_py` from the values in the dict as well?

jorisvandenbossche replied:
This is also what we seem to do on master

kszucs replied:
Updated, also extended the StructScalar's API to match a python mapping.

jorisvandenbossche commented on python/pyarrow/tests/test_scalars.py:
```diff
@@ -17,426 +17,395 @@
+
+def test_null_equality():
+    assert (pa.NA == pa.NA) is pa.NA
+    assert (pa.NA == 1) is pa.NA
```
I don't know to what extent we want to fully work out the scalars (so can certainly be a follow-up), but so the typed null scalars (not pa.NA), should probably behave the same as pa.NA, eg when it comes to equality (`pa.NA == 1` gives pa.NA, but `pa.scalar(None, pa.int64()) == 1` gives False)

kszucs replied:
Great question, I assume it should follow the same semantics as `Null` does.

kszucs commented:
> Really nice!
> 
> Could we add a `is_valid` attribute to the python scalar as well? Now the only way to check for a null value is to do `.as_py() is None` ?

Definitely!

> 
> Might not be related to the changes in this PR, but reviewing this triggered me to test scalar casting:
> 
> ```
> In [2]: s = pa.scalar(pd.Timestamp("2012-01-01")) 
>    ...:  
>    ...: import pyarrow.compute as pc 
>    ...: pc.cast(s, pa.timestamp('ns'))  
> ../src/arrow/compute/kernels/scalar_cast_temporal.cc:130:  Check failed: (batch[0].kind()) == (Datum::ARRAY) 
> ...
> Aborted (core dumped)
> ```
I think the cast kernel doesn't support scalars yet, on the other hand the scalars have custom `CastTo` implementation which we may want to remove in favor of `compute.Cast`?

cc @bkietz

kszucs commented on python/pyarrow/_dataset.pyx:
```diff
@@ -216,22 +216,18 @@ cdef class Expression:
-        elif isinstance(value, int):
-            scalar = MakeScalar(<int64_t>value)
-        elif isinstance(value, (bytes, str)):
-            scalar = MakeStringScalar(tobytes(value))
```
Removed it.

kszucs commented:
> Could we add a `is_valid` attribute to the python scalar as well? Now the only way to check for a null value is to do `.as_py() is None` ?

Added.

kszucs commented:
I'm considering to apply [cython.freelist](https://cython.readthedocs.io/en/latest/src/userguide/extension_types.html#fast-instantiation) on the scalar extension classes.

@wesm @pitrou @jorisvandenbossche do you have any positive experience with it?

I assume we should measure its impact before using it, so I can defer it to a follow-up.

pitrou commented on python/pyarrow/includes/libarrow.pxd:
```diff
@@ -44,6 +44,11 @@ cdef extern from "arrow/util/key_value_metadata.h" namespace "arrow" nogil:
 
+cdef extern from "arrow/util/decimal.h" namespace "arrow" nogil:
+    cdef cppclass CDecimal128" arrow::Decimal128":
+        c_string ToString(int32_t scale) const
```
Hmm, at some point we should be able to pretty-print scalars in C++.

kszucs replied:
Created jira https://issues.apache.org/jira/browse/ARROW-9279

pitrou commented on python/pyarrow/includes/libarrow.pxd:
```diff
@@ -877,20 +893,60 @@ cdef extern from "arrow/api.h" namespace "arrow" nogil:
         uint64_t value
 
+    cdef cppclass CHalfFloatScalar" arrow::HalfFloatScalar"(CScalar):
+        npy_half value
```
Ah, cool!

pitrou commented on python/pyarrow/lib.pxd:
```diff
@@ -179,101 +179,18 @@ cdef class Schema:
 
 
-cdef class FixedSizeBinaryValue(ArrayValue):
+cdef class NAType(Scalar):
```
`NullScalar` perhaps... also, why does it have its own class?

kszucs replied:
Forgot to rename it in the `.pxd` file.

pitrou commented on python/pyarrow/scalar.pxi:
```diff
@@ -16,1198 +16,745 @@
+        return self.wrapped
+
+    @property
+    def type(self):
```
Add docstrings here.

pitrou commented on python/pyarrow/scalar.pxi:
```diff
@@ -16,1198 +16,745 @@
-    cdef void _set_array(self, const shared_ptr[CArray]& sp_array):
-        self.sp_array = sp_array
+    def __eq__(self, other):
+        return NA
```
Hmm, what?

kszucs replied:
Well, it was the behavior before the refactor. Shall we check proper equality with NullScalar instead?

pitrou replied:
Well, it should probably return `False` for non-nulls?

kszucs replied:
Agree with that, just worried a bit about backward compatibility.

kszucs replied:
Updated to return False.

jorisvandenbossche replied:
> Updated to return False.

Although this might actually get into the discussion .., but if being "conversative" on this PR pending the mailing list discussion, I would rather simply remove the override here and use the base class `__eq__` which relies on C++'s Equals (then at least all scalar classes are consistent on this, and not only NullScalar deviating from the others).  
Because changing it to False both 1) breaks with the current behaviour as 2) breaks with the `equals` method / deviates from the other scalar classes (and it seems likely that we want to do one of both)

jorisvandenbossche replied:
Ah, looking at the actual diff of your latest changes, you didn't update to return False, but to return True (for comparing with another Null value, False for comparing with any other value of course, but I had the first case in my mind when reacting ;))

So that basically does what I proposed above, but, I think we can still remove the `__eq__` alltogether to use the base class implementation, which should give the same behaviour.

kszucs replied:
Indeed, the c++ impl should return the same.

pitrou commented on python/pyarrow/scalar.pxi:
```diff
@@ -16,1198 +16,745 @@
-            return repr(self.as_py())
-        else:
-            return super(Scalar, self).__repr__()
+    def __hash__(self):
```
Why do you need to redefine this method?

kszucs replied:
Once I override `__eq__` the inherited `__hash__` gets removed.

pitrou commented on python/pyarrow/scalar.pxi:
```diff
@@ -16,1198 +16,745 @@
+    Concrete class for float scalars.
     """
 
+    def __hash__(self):
```
Why do you need to redefine this?

kszucs replied:
Same as above.

pitrou commented on python/pyarrow/scalar.pxi:
```diff
@@ -16,1198 +16,745 @@
+        return hasher(self.wrapped)
+
+    def __eq__(self, other):
+        if hasattr(self, 'as_py'):
```
This `hasattr` check is pointless as, `self` is a `HalfFloatScalar`.

kszucs replied:
Right.

pitrou commented on python/pyarrow/scalar.pxi:
```diff
@@ -16,1198 +16,745 @@
+            CDecimal128Scalar* sp = <CDecimal128Scalar*> self.wrapped.get()
+            CDecimal128Type* dtype = <CDecimal128Type*> sp.type.get()
+        if sp.is_valid:
+            return _pydecimal.Decimal(
```
Why `_pydecimal`? You should be using the `decimal` module.

kszucs replied:
`decimal` is imported as `_pydecimal` in [lib.pyx](https://github.com/apache/arrow/blob/master/python/pyarrow/lib.pyx#L23)

pitrou replied:
Wow, this is misleading, before `_pydecimal` is the name of the pure Python implementation of `decimal`:
```python
>>> import _pydecimal                                                                                                                                                          
>>> _pydecimal                                                                                                                                                                 
<module '_pydecimal' from '/home/antoine/miniconda3/envs/pyarrow/lib/python3.7/_pydecimal.py'>
>>> import decimal                                                                                                                                                             
>>> decimal is _pydecimal                                                                                                                                                      
False
```

pitrou commented on python/pyarrow/scalar.pxi:
```diff
@@ -16,1198 +16,745 @@
         if not dtype.timezone().empty():
             tzinfo = string_to_tzinfo(frombytes(dtype.timezone()))
+            if not isinstance(tzinfo, datetime.tzinfo):
+                tzinfo = string_to_tzinfo(tzinfo)
```
Why do you have to do this twice? If `string_to_tzinfo` is broken, it should be fixed IMO.

kszucs replied:
It's a copy issue, I incrementally ported the previous helper functions. Removing the second call.

pitrou commented on python/pyarrow/scalar.pxi:
```diff
@@ -16,1198 +16,745 @@
-        return ap.Value(self.index)
+        buffer = self.as_buffer()
+        if buffer is not None:
+            return self.as_buffer().to_pybytes()
```
`buffer.to_pybytes()`

pitrou commented on python/pyarrow/scalar.pxi:
```diff
@@ -16,1198 +16,745 @@
-        cdef CDoubleArray* ap = <CDoubleArray*> self.sp_array.get()
-        return ap.Value(self.index)
+
+cdef class FixedSizeBinaryScalar(BinaryScalar):
```
Same here.

pitrou commented on python/pyarrow/scalar.pxi:
```diff
@@ -16,1198 +16,745 @@
+            return None
+
 
+cdef class LargeStringScalar(StringScalar):
```
Same here.

pitrou commented on python/pyarrow/scalar.pxi:
```diff
@@ -16,1198 +16,745 @@
-    def as_py(self):
-        """
-        Return this value as a Python string.
+        The exact type depends on the dictionary value type.
```
Not "dictionary"

pitrou commented on python/pyarrow/scalar.pxi:
```diff
@@ -1217,21 +764,50 @@ cdef dict _scalar_classes = {
-        value = klass.__new__(klass)
-        value.init(type, sp_array, index)
-        return value
+def scalar(value, DataType type=None, bint safe=True,
```
Can you add a docstring? Also, make `safe` and `memory_pool` keyword-only.

pitrou replied:
It seems `safe` isn't used btw.

kszucs replied:
Removed.

pitrou commented on python/pyarrow/scalar.pxi:
```diff
@@ -1217,21 +764,50 @@ cdef dict _scalar_classes = {
+
+    options.size = 1
+    options.pool = maybe_unbox_memory_pool(memory_pool)
+    # options.from_pandas = from_pandas
```
What?

kszucs replied:
Implemented it.

pitrou commented on python/pyarrow/tests/test_convert_builtin.py:
```diff
@@ -968,25 +968,31 @@ def test_sequence_timestamp_from_int_with_unit():
     assert len(arr_s) == 1
     assert arr_s.type == s
-    assert repr(arr_s[0]) == "datetime.datetime(1970, 1, 1, 0, 0, 1)"
+    assert repr(arr_s[0].as_py()) == "datetime.datetime(1970, 1, 1, 0, 0, 1)"
```
What is being tested here? If you're testing the `as_py()` result, then `repr` isn't needed. Otherwise, you may instead call `str(arr_s[0])`.

kszucs replied:
Updated.

pitrou commented on python/pyarrow/tests/test_dataset.py:
```diff
@@ -385,10 +385,12 @@ def test_partitioning():
 
 def test_expression_serialization():
     a = ds.scalar(1)
+    a_ = ds.scalar(pa.scalar(1))
```
Uh, can you use another name? Why not `g`?

kszucs replied:
Updated.

pitrou commented on python/pyarrow/tests/test_parquet.py:
```diff
@@ -2028,7 +2028,7 @@ def test_filters_invalid_pred_op(tempdir, use_legacy_dataset):
 
-    with pytest.raises(ValueError if use_legacy_dataset else TypeError):
+    with pytest.raises(ValueError):
         # dataset API returns TypeError when trying create invalid comparison
```
Comment seems obsolete?

pitrou commented on python/pyarrow/tests/test_scalars.py:
```diff
@@ -16,427 +16,443 @@
+    assert hash(s) == hash(s)
+    assert s.is_valid is True
+    with pytest.warns(FutureWarning):
+        isinstance(s, deprecated)
```
`assert isinstance` perhaps?

pitrou commented on python/pyarrow/tests/test_scalars.py:
```diff
@@ -16,427 +16,443 @@
+    assert s == value
+    assert s == s
+    assert s != "else"
+    assert hash(s) == hash(s)
```
What's the point of this test? This is going to be trivially true.

kszucs replied:
It tests that different calls of the c++ hashing mechanism produces the same hash values.

pitrou commented on python/pyarrow/tests/test_scalars.py:
```diff
@@ -16,427 +16,443 @@
+    s = pa.scalar(value, type=ty)
+    assert isinstance(s, klass)
+    assert s == value
+    assert s == s
```
It would be more interesting to instantiate `s` a second time, e.g. `assert s == pa.scalar(value, type=ty)`

kszucs replied:
Updated.

pitrou commented on python/pyarrow/tests/test_scalars.py:
```diff
@@ -16,427 +16,443 @@
+
+    s = pa.scalar(None, type=s.type)
+    assert s.is_valid is False
+    assert s.as_py() is None
```
Also `assert s != pa.scalar(value, type=ty)`

kszucs replied:
Updated.

pitrou commented on python/pyarrow/tests/test_scalars.py:
```diff
@@ -16,427 +16,443 @@
+
+
+def test_null_singleton():
+    with pytest.raises(Exception):
```
You should a test more precise exception.

kszucs replied:
Changed it to RuntimeException.

pitrou commented on python/pyarrow/tests/test_scalars.py:
```diff
@@ -16,427 +16,443 @@
+    null = pa.scalar(None)
+    assert null is pa.NA
+    assert null.as_py() is None
+    assert (null == "something") is pa.NA
```
This looks incorrect and should return `False`.

pitrou replied:
Also `assert null == pa.scalar(None)`

kszucs replied:
Added.

pitrou commented:
Ok, at a quick glance, it seems that null container tests work properly regardless:
```python
>>> s = set()                                                                                                                                                                  
>>> s.add(pa.scalar(None))                                                                                                                                                     
>>> s                                                                                                                                                                          
{<pyarrow.NullScalar: None>}
>>> pa.scalar(None) in s                                                                                                                                                       
True
>>> s.add(pa.scalar(None, pa.int64()))                                                                                                                                         
>>> s.add(pa.scalar(12, pa.int64()))                                                                                                                                           
>>> s                                                                                                                                                                          
{<pyarrow.Int64Scalar: 12>,
 <pyarrow.NullScalar: None>,
 <pyarrow.Int64Scalar: None>}
>>> pa.scalar(None, pa.int64()) in s                                                                                                                                           
True
>>> pa.scalar(None, pa.int32()) in s                                                                                                                                           
False
```
```python
>>> l = [pa.scalar(None)]                                                                                                                                                      
>>> pa.scalar(None) in l                                                                                                                                                       
True
>>> pa.scalar(None, pa.int64()) in l                                                                                                                                           
False
```

kszucs commented on python/pyarrow/scalar.pxi:
```diff
@@ -16,1198 +16,745 @@
+    # @property
+    # def index(self):
+    #     """
+    #     Return this value's underlying index as a scalar.
```
Waited for Ben's update from https://github.com/apache/arrow/pull/7536

jorisvandenbossche commented:
> the null equality tests look like a nuisance for regular Python usage (`__eq__` should return a boolean)

The reason that those scalars return null on equality check and not True/False, it to ensure consistent behaviour between arrays and scalars (to ensure `(arr1 == arr2)[0]` and `arr1[0] == arr2[0]`gives the same answer). 
Our element-wise equality propagates null, and so IMO also the scalar should do.

pitrou commented:
The problem is that it breaks Python semantics in potentially annoying places:
```python
>>> import pyarrow as pa                                                                                                                                                 
>>> na = pa.scalar(None)                                                                                                                                                 
>>> na in [5]                                                                                                                                                            
True
```

jorisvandenbossche commented:
We could "fix" that one by raising in `__bool__` (meaning: it will at least give an error instead of silently returning a wrong answer)

pitrou commented:
Yes, we could. That may have other annoying implications, though (such as `__contains__` not working anymore). I've started a ML discussion.

jorisvandenbossche commented on python/pyarrow/scalar.pxi:
```diff
@@ -16,1198 +16,748 @@
+        return str(self.as_py())
+
+    def equals(self, Scalar other):
+        return self.wrapped.get().Equals(other.unwrap().get()[0])
```
Using the C++ `Equals` for `==` might give some potentially "surprising" behaviours, eg `pa.scalar(1, type=pa.int64()) == pa.scalar(1, type=pa.int32())` gives False (while `pa.scalar(1, type=pa.int64()) == 1` or `pa.scalar(1, type=pa.int64()) == np.int32(1)` gives True).

jorisvandenbossche replied:
Although, also for arrays this currently doesn't work. But for arrays it gives an error about no kernel matching the types (which is more informative than "silently" giving False)

kszucs replied:
Removed the options `.as_py()` conversion from the equality check, so now a scalar can be equal to another scalar only.

jorisvandenbossche replied:
> Removed the options .as_py() conversion from the equality check, so now a scalar can be equal to another scalar only.

Sounds good!

jorisvandenbossche commented on python/pyarrow/scalar.pxi:
```diff
@@ -1217,21 +767,95 @@ cdef dict _scalar_classes = {
+    -----
+    Localized timestamps will currently be returned as UTC (pandas's native
+    representation). Timezone-naive data will be implicitly interpreted as
+    UTC.
```
I don't think this is fully correct? (or at least a bit misleading/confusing) 
Timezone-naive data simply have no timezone, I am not aware of places in arrow where we implicitly assume UTC?

kszucs replied:
Removing. Copied from `pa.array()`'s docstring, so we mey need to update it there as well.

kszucs commented on python/pyarrow/scalar.pxi:
```diff
@@ -16,1198 +16,748 @@
+                      "please use the `value` property instead",
+                      FutureWarning)
+        index = self.index
+        return None if index is None else self.index
```
Updated.

kszucs commented on python/pyarrow/scalar.pxi:
```diff
@@ -16,1198 +16,748 @@
+        warnings.warn("`dictionary_value` property is deprecated as of 1.0.0, "
+                      "please use the `value` property instead", FutureWarning)
+        value = self.value
+        return None if value is None else self.value
```
Updated.

kszucs commented on python/pyarrow/scalar.pxi:
```diff
@@ -16,1198 +16,748 @@
+    def index_value(self):
+        warnings.warn("`dictionary_value` property is deprecated as of 1.0.0"
+                      "please use the `value` property instead",
+                      FutureWarning)
```
Updated.

kszucs commented on python/pyarrow/scalar.pxi:
```diff
@@ -16,1198 +16,745 @@
-    """
-    Concrete class for float64 array elements.
-    """
+cdef class LargeBinaryScalar(BinaryScalar):
```
Updated it to cast to`CBaseBinaryScalar` which is the common root of the binary-like scalar types.

kszucs commented on python/pyarrow/scalar.pxi:
```diff
@@ -16,1198 +16,745 @@
-        buf = SliceBuffer(ap.value_data(), ap.value_offset(self.index),
-                          ap.value_length(self.index))
-        return pyarrow_wrap_buffer(buf)
+cdef class FixedSizeListScalar(ListScalar):
```
Changed to use the common root `CBaseListScalar`.

kszucs commented:
Addressed the review comments and changed the equality checks to be strict, only allowing to compare scalars with scalars of the same type. One exception *could* be to compare invalid scalar values to `pa.NA`, but even if we consider to choose that behavior, it should be implemented on the C++ side.

@pitrou @jorisvandenbossche it should be ready for another review.

jorisvandenbossche commented on python/pyarrow/tests/test_array.py:
```diff
@@ -583,12 +584,14 @@ def test_dictionary_from_numpy():
+        # if mask[i]:
+        #     assert d2[i].as_py() is None
+        # else:
+        #     assert d2[i].as_py() == dictionary[indices[i]]
```
left-over ?

kszucs replied:
Fixed.

kszucs commented:
@pitrou added the C++ tests.

pitrou commented:
Looks like the PR needs rebasing and fixing for the latest union changes.

pitrou commented on cpp/src/arrow/scalar.cc:
```diff
@@ -185,6 +192,35 @@ DictionaryScalar::DictionaryScalar(std::shared_ptr<DataType> type)
+      index_value = checked_cast<const Int64Scalar&>(*value.index).value;
+      break;
+    default:
+      return Status::Invalid("Not implemented dictionary index type");
```
`Status::TypeError`?

kszucs replied:
Updating.

pitrou commented on cpp/src/arrow/scalar_test.cc:
```diff
@@ -80,6 +84,15 @@ TYPED_TEST(TestNumericScalar, Basics) {
+  ASSERT_OK_AND_ASSIGN(auto two, arr->GetScalar(2));
+  ASSERT_TRUE(null->Equals(*null_value));
+  ASSERT_TRUE(one->Equals(ScalarType(1)));
+  ASSERT_TRUE(two->Equals(ScalarType(2)));
```
You never check that `Equals` returns false?

kszucs replied:
Not yet, adding check for it.

pitrou commented on cpp/src/arrow/scalar_test.cc:
```diff
@@ -451,6 +565,127 @@ TEST(TestStructScalar, FieldAccess) {
+  ASSERT_TRUE(third->Equals(scalar_beta));
+
+  ASSERT_OK_AND_ASSIGN(auto fourth, arr.GetScalar(3));
+  ASSERT_TRUE(fourth->Equals(MakeNullScalar(ty)));
```
Hmm... interesting. Shouldn't it be `MakeNullScalar(utf8())`?

kszucs replied:
I think `GetScalar()` should return a Scalar with the same type as the Array.

pitrou replied:
Ah, I see. Yes, you're probably right. A pity we lose information about the value type, then.

kszucs replied:
We have the value type in `UnionScalar.value.type`.

pitrou replied:
Ah, perhaps check that in the test?

kszucs replied:
Adding.

kszucs replied:
So for UnionScalar we have two `is_valid` flags, une for the union scalar and one for the underlying value scalar. Currently `MakeNullScalar()` sets the validity flag for the union scalar and leaves the value scalar uninitialized. 

What do you think, shall we initialize the value scalar as well?

pitrou replied:
It the value scalar is not a nullptr, then it would seem better, unless that's complicated.

kszucs replied:
We can ensure to have an explicit type to the value scalar from `array.GetScalar()` but not when a null union scalar is directly constructed, so `MakeNullScalar(union_type)` won't know which type to choose for the underlying value scalar.

pitrou commented on cpp/src/arrow/scalar_test.cc:
```diff
@@ -451,6 +565,127 @@ TEST(TestStructScalar, FieldAccess) {
+  ASSERT_TRUE(fourth->Equals(MakeNullScalar(ty)));
+
+  ASSERT_OK_AND_ASSIGN(auto fifth, arr.GetScalar(4));
+  ASSERT_TRUE(fifth->Equals(MakeNullScalar(ty)));
```
Similarly, shouldn't it be `MakeNullScalar(uint64())`?

pitrou commented on cpp/src/arrow/scalar_test.cc:
```diff
@@ -451,6 +565,127 @@ TEST(TestStructScalar, FieldAccess) {
+  ASSERT_TRUE(third->Equals(scalar_beta));
+
+  ASSERT_OK_AND_ASSIGN(auto fourth, arr.GetScalar(3));
+  ASSERT_TRUE(fourth->Equals(MakeNullScalar(ty)));
```
`MakeNullScalar(utf8)`?

wesm commented:
+1. Thanks @kszucs for this work!
